source/base/tensor_product_polynomials_bubbles.cc

```diff
@@ -0,0 +1,251 @@
+                else
+                  derivative=1;
+              }
+            grad_grad_1[d1][d2] *= v[x][derivative];
+          }
+      }
+
+  //calculate (\partial_j  \psi) *(\partial_k monomial)
+  // and (\partial_k  \psi) *(\partial_j monomial)
+  Tensor<2,dim> grad_grad_2;
+  Tensor<2,dim> grad_grad_3;
+  for (unsigned int d=0; d<dim; ++d)
+    {
+      grad_grad_2[d][comp] = v[dim][1];
+      grad_grad_3[comp][d] = v[dim][1];
+      for (unsigned int x=0; x<dim; ++x)
+        {
+          grad_grad_2[d][comp] *= v[x][d==x];
+          grad_grad_3[comp][d] *= v[x][d==x];
+        }
+    }
+
+  //calculate \psi *(\partial j \partial_k monomial) and sum
+  Tensor<2,dim> grad_grad;
+  double psi_value = 1.;
+  for (unsigned int x=0; x<dim; ++x)
+    psi_value *= v[x][0];
+
+  for (unsigned int d1=0; d1<dim; ++d1)
+    for (unsigned int d2=0; d2<dim; ++d2)
+      grad_grad[d1][d2] = grad_grad_1[d1][d2]
+                          +grad_grad_2[d1][d2]
+                          +grad_grad_3[d1][d2];
+  grad_grad[comp][comp]+=psi_value*v[dim][2];
+
+  return grad_grad;
+}
+
+template <int dim>
+void
+TensorProductPolynomialsBubbles<dim>::
+compute (const Point<dim>            &p,
+         std::vector<double>         &values,
+         std::vector<Tensor<1,dim> > &grads,
+         std::vector<Tensor<2,dim> > &grad_grads) const
+{
+  const unsigned int q_degree = this->polynomials.size()-1;
+  const unsigned int max_indices = this->n_tensor_pols;
+  const unsigned int n_bubbles = ((q_degree<=1)?1:dim);
+  Assert (values.size()==max_indices+n_bubbles || values.size()==0,
+          ExcDimensionMismatch2(values.size(), max_indices+n_bubbles, 0));
+  Assert (grads.size()==max_indices+n_bubbles     || grads.size()==0,
+          ExcDimensionMismatch2(grads.size(), max_indices+n_bubbles, 0));
+  Assert (grad_grads.size()==max_indices+n_bubbles || grad_grads.size()==0,
+          ExcDimensionMismatch2(grad_grads.size(), max_indices+n_bubbles, 0));
+
+  bool do_values = false, do_grads = false, do_grad_grads = false;
+  unsigned int n_values_and_derivatives;
+  if (values.empty() == false)
+    {
+      values.resize(this->n_tensor_pols);
+      do_values = true;
+      n_values_and_derivatives = 1;
+    }
+  if (grads.empty() == false)
+    {
+      grads.resize(this->n_tensor_pols);
+      do_grads = true;
+      n_values_and_derivatives = 2;
+    }
+  if (grad_grads.empty() == false)
+    {
+      grad_grads.resize(this->n_tensor_pols);
+      do_grad_grads = true;
+      n_values_and_derivatives = 3;
+    }
+
+  this->TensorProductPolynomials<dim>::compute(p, values, grads, grad_grads);
+
+  //TODO: Do this more efficiently!
+  for (unsigned int i=this->n_tensor_pols; i<this->n_tensor_pols+n_bubbles; ++i)
+    {
+      if (do_values)
+        values.push_back(compute_value(i,p));
+      if (do_grads)
+        grads.push_back(compute_grad(i,p));
+      if (do_grad_grads)
+        grad_grads.push_back(compute_grad_grad(i,p));
+    }
+}
```

Comment:
Given that the formulas behind the functions you call here do quite some math I'd assume the additional function call overhead to be minimal and can't imagine a noticeable gain. So I'd suggest you remove the TODO here.